### ISSUE TITLE
Can't update to V0.42 - ResolutionRequest: conversion failed to version v1beta1

### ISSUE THREAD

**Issue by ArturFortunato** [tektoncd/pipeline#5873]
# Expected Behavior

I can run pipelines in version 0.42

# Actual Behavior

The resources are deployed, but my pipelineRuns are stuck in "Pending" state. My tekton-pipelines-webhook logs show the following:

```json
{"severity":"error","timestamp":"2022-12-14T10:48:14.017Z","logger":"tekton-pipelines-webhook","caller":"conversion/conversion.go:59","message":"Conversion failed","commit":"c3c07aa","uid":"9d8ddab9-aa14-4517-acf5-67955e3e1ca8","desiredAPIVersion":"resolution.tekton.dev/v1beta1","error":"conversion failed to version v1beta1 for type [kind=ResolutionRequest group=resolution.tekton.dev version=v1alpha1] -  v1beta1 is the highest known version, got: *v1alpha1.ResolutionRequest","stacktrace":"knative.dev/pkg/webhook/resourcesemantics/conversion.(*reconciler).Convert\n\tknative.dev/pkg@v0.0.0-20221011175852-714b7630a836/webhook/resourcesemantics/conversion/conversion.go:59\nknative.dev/pkg/webhook.conversionHandler.func1\n\tknative.dev/pkg@v0.0.0-20221011175852-714b7630a836/webhook/conversion.go:66\nnet/http.HandlerFunc.ServeHTTP\n\tnet/http/server.go:2084\nnet/http.(*ServeMux).ServeHTTP\n\tnet/http/server.go:2462\nknative.dev/pkg/webhook.(*Webhook).ServeHTTP\n\tknative.dev/pkg@v0.0.0-20221011175852-714b7630a836/webhook/webhook.go:262\nknative.dev/pkg/network/handlers.(*Drainer).ServeHTTP\n\tknative.dev/pkg@v0.0.0-20221011175852-714b7630a836/network/handlers/drain.go:113\nnet/http.serverHandler.ServeHTTP\n\tnet/http/server.go:2916\nnet/http.(*conn).serve\n\tnet/http/server.go:1966"}
```

# Steps to Reproduce the Problem

1. Deploy version 0.42
2. Create a sample pipeline
3. Create a pipelineRun with it
4. Check the `tekton-pipelines-webhook` logs

# Additional Info

- Kubernetes version:

  **Output of `kubectl version`:**

Client Version: version.Info{Major:"1", Minor:"25", GitVersion:"v1.25.0", GitCommit:"a866cbe2e5bbaa01cfd5e969aa3e033f3282a8a2", GitTreeState:"clean", BuildDate:"2022-08-23T17:44:59Z", GoVersion:"go1.19", Compiler:"gc", Platform:"darwin/arm64"}
Server Version: version.Info{Major:"1", Minor:"23+", GitVersion:"v1.23.13-eks-fb459a0", GitCommit:"55bd5d5cb7d32bc35e4e050f536181196fb8c6f7", GitTreeState:"clean", BuildDate:"2022-10-24T20:35:40Z", GoVersion:"go1.17.13", Compiler:"gc", Platform:"linux/amd64"}

- Tekton Pipeline version:

  **Output of `tkn version` or `kubectl get pods -n tekton-pipelines -l app=tekton-pipelines-controller -o=jsonpath='{.items[0].metadata.labels.version}'`**

0.42

**Comment by ArturFortunato** [comment]
Sorry, missed the previous version.

I tried upgrading from 0.40 to 0.42 and it failed. I then tried to upgrade from 0.41, which failed as well, and then from 0.41 to 0.42, which again didn't work (leading to the same result as 0.40 directly to 0.42).

Regarding Pipeline, here it is

```
apiVersion: tekton.dev/v1beta1
kind: Pipeline
metadata:
  name: pipeline
  namespace: tekton-pipelines
  annotations:
    tekton.dev/git-0: https://github.com
    kitch/serviceAccountName: tekton
spec:
  params:
    - name: branch
      description: PR base branch
      type: string
    - name: pull_request_number
      description: The id of the associated pull request
    - name: git_revision
      description: Unique commit identification
  resources:
    - name: source-repo
      type: git
  workspaces:
    - name: shared
    - name: output
  tasks:
    - name: get-changed-applications
      resources:
        inputs:
          - name: source-repo
            resource: source-repo
      workspaces:
        - name: shared
          workspace: shared
      params:
      - name: applications_folder
        value: applications
      - name: applications_file
        value: applications.txt
      - name: github_secret_name
        value: tekton-pipelines
      - name: git_revision
        value: $(params.git_revision)
      - name: branch
        value: $(params.branch)
      taskSpec:
        volumes:
          - name: github-secret
            secret:
              secretName: $(params.github_secret_name)
        params:
          - name: applications_file
            type: string
            description: "Name of the file to store application names. Defaults to 'applications.txt'"
          - name: applications_folder
            type: string
            description: "Name of the folder to store the applications names file (will be created inside the shared workspace"
          - name: github_secret_name
            type: string
          - name: git_revision
            type: string
          - name: branch
            type: string
        resources:
          inputs:
            - name: source-repo
              type: git
        workspaces:
          - name: shared
        steps:
        - name: list-changed-applications-names
          image: alpine/git
          # run as root to have permissions to write in the shared workspace
          securityContext:
            runAsUser: 0
          script: |
            #!/bin/sh

            ...
       
        - name: clone-changed-applications
          image: alpine/git
          # run as root to have permissions to write in the shared workspace
          securityContext:
            runAsUser: 0
          env:
            - name: GH_USER
              valueFrom:
                secretKeyRef:
                  name: tekton-pipelines
                  key: username
            - name: GH_TOKEN
              valueFrom:
                secretKeyRef:
                  name: tekton-pipelines
                  key: personal-access-token
          script: |
            #!/bin/sh

            ...

    - name: create-and-update-applications
      when:
        - input: $(params.branch)
          operator: in
          values:
            - master
      runAfter:
        - get-changed-applications
      resources:
        inputs:
          - name: source-repo
            resource: source-repo
      workspaces:
        - name: shared
          workspace: shared
      params:
      - name: applications_folder
        value: applications
      - name: applications_file
        value: applications.txt
      - name: github_org
        value: "XXX"
      - name: github_user
        value: "XXXX"
      - name: github_secret_name
        value: tekton-pipelines
      - name: github_access_token_key
        value: XXX
      - name: external_domain
        value: XXX
      taskSpec:
        volumes:
          - name: github-secret
            secret:
              secretName: $(params.github_secret_name)
        params:
          - name: applications_file
            type: string
            description: "Name of the file to store application names. Defaults to 'applications.txt'"
          - name: applications_folder
            type: string
            description: "Name of the folder to store the applications names file (will be created inside the shared workspace"
          - name: webhook_events
            description: "List of events the webhook will send notifications for"
            default: '[\"push\",\"pull_request\"]'
          - name: github_org
            type: string
          - name: github_user
            type: string
          - name: github_secret_name
            type: string
          - name: github_access_token_key
            type: string
          - name: external_domain
            type: string
        resources:
          inputs:
            - name: source-repo
              type: git
        workspaces:
          - name: shared
        steps:
        - name: create-application-webhooks
          image: curlimages/curl:latest
          volumeMounts:
          - name: github-secret
            mountPath: /var/secret
          script: |
            #!/bin/sh

          ...

        - name: apply-pipelines
          image: XXX
          script: |
            #!/bin/sh

           .....

    - name: clone-hcl
      runAfter:
        - get-changed-applications
      when:
        - input: $(params.branch)
          operator: notin
          values:
            - master
      workspaces:
        - name: shared
          workspace: shared
      params:
        - name: repository_name
          value: XXX
        - name: github_secret_name
          value: tekton-pipelines
      taskRef:
        name: git-clone

    - name: clone-blueprints
      runAfter:
        - get-changed-applications
      when:
        - input: $(params.branch)
          operator: notin
          values:
            - master
      workspaces:
        - name: shared
          workspace: shared
      params:
        - name: repository_name
          value: XXX
        - name: github_secret_name
          value: tekton-pipelines
      taskRef:
        name: git-clone
    
    - name: sync-with-hcl-remote-branch
      runAfter:
        - clone-hcl
      workspaces:
        - name: shared
          workspace: shared
      params:
        - name: branch
          value: $(params.branch)
      taskRef:
        name: sync-with-hcl-remote-branch

    - name: generate-hcl
      runAfter:
        - clone-blueprints
        - sync-with-hcl-remote-branch
      when:
        - input: $(params.branch)
          operator: notin
          values:
            - master
      resources:
        inputs:
          - name: source-repo
            resource: source-repo
      workspaces:
        - name: shared
          workspace: shared
      params:
      - name: applications_folder
        value: applications
      - name: applications_file
        value: applications.txt
      taskSpec:
        volumes:
          - name: github-secret
            secret:
              secretName: $(params.github_secret_name)
        params:
          - name: applications_file
            type: string
            description: "Name of the file to store application names."
          - name: applications_folder
            type: string
            description: "Name of the folder to store the applications names file (will be created inside the shared workspace"
        resources:
          inputs:
            - name: source-repo
              type: git
        workspaces:
          - name: shared
        steps:
        - name: generate-scripts
          image: alpine
          script: |
            #!/bin/sh

            ...

        - name: generate-hcl
          image: XXX
          # run as root to have permissions to write in the shared workspace
          securityContext:
            runAsUser: 0
          script: |
            #!/bin/sh

            ...
    - name: open-or-update-hcl-pull-request
      when:
        - input: $(params.branch)
          operator: notin
          values:
            - master
      runAfter:
        - generate-hcl
      taskRef:
        name: open-or-update-hcl-pull-request
      params:
        - name: pull_request_number
          value: $(params.pull_request_number)
        - name: branch
          value: $(params.branch)
        - name: github_secret_name
          value: tekton-pipelines
        - name: origin_repository
          value: kitch-application-catalog
      workspaces:
        - name: shared
          workspace: shared
      resources:
        inputs:
          - name: source-repo
            resource: source-repo
```

**Comment by vdemeester** [comment]
@ArturFortunato how does the `PipelineRun` looks like ? Given the error on the `ResolutionRequest` type, this might mean you use remote resolution.

**Comment by XinruZhang** [comment]
Thanks for reporting the issue! @ArturFortunato!

**TL;DR** 

To address the issue, you can reinstall `v0.40.0`, run `kubectl delete resolutionrequests --all` to delete all `v1alpha1.ResolutionRequest`, then install `v0.42.0`, everything would be fine.

---

The reason: For `ResolutionRequest`, we don't convert from `v1alpha1` to `v1beta1`, vice versa. Thus, even though we "serve" the `v1alpha1`, it actually doesn't work.

https://github.com/tektoncd/pipeline/blob/c39fd32a80706d2ea5af0bab5fe87b5cffea36ae/config/300-resolutionrequest.yaml#L36

**How to reproduce the error:**

**1. Install, Tekton `v0.40.0` with a cluster resolver, create a PipelineRun that use remote resolution**

1.1. Install Tekton `v0.40.0` with a cluster resolver installed
1.2. create a `Task` called `simple-task`:
```yaml
apiVersion: tekton.dev/v1beta1
kind: Task
metadata:
  name: simple-task
spec:
  steps:
    - image: busybox
      args: ['-c', 'sleep 10']
```
1.3. deploy a `PipelineRun` that **refers to `simple-task`**
```yaml
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: pr-remote-task
spec:
  pipelineSpec:
    tasks:
    - name: pr-simple-task
      taskRef:
        resolver: cluster
        params:
        - name: kind
          value: task
        - name: name
          value: simple-task
        - name: namespace
          value: default
```

At this time, we'll have an `v1alpha1. ResolutionRequest` created
```shell
kubectl  get resolutionrequests
NAME                                       SUCCEEDED   REASON
cluster-d6c1b1cb008cbb8666ee478eba9e496a   True
```

**2. Install, Tekton `v0.42.0` create a PipelineRun that use remote resolution**

This step will produce the error this issue reports: PipelineRun is stuck, and the webhook controller logs out the same error mentioned in the description.

**Comment by XinruZhang** [comment]
@abayer  @jerop 

Considering we don't have the conversion between v1alpha1 and v1beta1. Do we still want to keep serving the alpha version?

**Comment by abayer** [comment]
I thought we had conversion… I’ll take a look tomorrow.

**Comment by abayer** [comment]
We do, but I think I screwed up https://github.com/tektoncd/pipeline/blob/main/cmd/webhook/main.go#L195.

**Comment by XinruZhang** [comment]
Npnp, let's fix it! But indeed, we need an automated way to capture this beforehand instead of relying on human knowledge, this can be handled by @JeromeJu's work on upgrade test: case ii of https://github.com/tektoncd/pipeline/issues/5782, originated from @lbernick's comment: https://github.com/tektoncd/pipeline/issues/5193#issue-1313586664

**Comment by ArturFortunato** [comment]
> @ArturFortunato how does the `PipelineRun` looks like ? Given the error on the `ResolutionRequest` type, this might mean you use remote resolution.

No, the Pipeline resource is defined in the same cluster, so I don't think so

**Comment by ArturFortunato** [comment]
> Thanks for reporting the issue! @ArturFortunato!
> 
> **TL;DR**
> 
> To address the issue, you can
> 
> 1. uninstall `v0.42.0`
> 2. reinstall `v0.40.0`
> 3. run `kubectl delete resolutionrequests --all` to delete all `v1alpha1.ResolutionRequest`
> 4. install `v0.42.0` again
> 
> Everything would be fine.
> 
> **What's the error?**
> 
> For `ResolutionRequest`, we don't convert from `v1alpha1` to `v1beta1`, vice versa. Thus, even though we "serve" the `v1alpha1`, it actually doesn't work.
> 
> https://github.com/tektoncd/pipeline/blob/c39fd32a80706d2ea5af0bab5fe87b5cffea36ae/config/300-resolutionrequest.yaml#L36
> 
> **How to reproduce the error?**
> 
> **1. Install, Tekton `v0.40.0` with a cluster resolver, create a PipelineRun that use remote resolution**
> 
> * `kubectl edit configmap resolvers-feature-flags -n tekton-pipelines-resolvers`, set `enable-cluster-resolver` as `true`.
> * `kubectl edit configmap feature-flags -n tekton-pipelines`, set `enable-api-fields` as `v1alpha1`
> * create a `PipelineRun` that resolving an in-cluster task:
> 
> ```yaml
> apiVersion: tekton.dev/v1beta1
> kind: Task
> metadata:
>   name: simple-task-alpha-rr
> spec:
>   steps:
>     - image: busybox
>       args: ['-c', 'sleep 10']
> ---
> apiVersion: tekton.dev/v1beta1
> kind: PipelineRun
> metadata:
>   name: pr-remote-task-alpha-rr
> spec:
>   pipelineSpec:
>     tasks:
>     - name: pr-simple-task-alpha-rr
>       taskRef:
>         resolver: cluster
>         params:
>         - name: kind
>           value: task
>         - name: name
>           value: simple-task-alpha-rr
>         - name: namespace
>           value: default
> ```
> 
> At this time, we'll have an `v1alpha1. ResolutionRequest` created
> 
> ```shell
> kubectl  get resolutionrequests
> NAME                                       SUCCEEDED   REASON
> cluster-d6c1b1cb008cbb8666ee478eba9e496a   True
> ```
> 
> **2. Install Tekton `v0.42.0`, then create a PipelineRun that uses remote resolution**
> 
> ```yaml
> apiVersion: tekton.dev/v1beta1
> kind: Task
> metadata:
>   name: simple-task
> spec:
>   steps:
>     - image: busybox
>       args: ['-c', 'sleep 10']
> ---
> apiVersion: tekton.dev/v1beta1
> kind: PipelineRun
> metadata:
>   name: pr-remote-task
> spec:
>   pipelineSpec:
>     tasks:
>     - name: pr-simple-task
>       taskRef:
>         resolver: cluster
>         params:
>         - name: kind
>           value: task
>         - name: name
>           value: simple-task
>         - name: namespace
>           value: default
> ```
> 
> This step will produce the error this issue reports: PipelineRun is stuck, and the webhook controller logs out the same error mentioned in the description.

Thank you, works like a charm 🥳 🥳
